firedrake/assign.py

```diff
@@ -0,0 +1,251 @@
+        else:
+            raise ValueError("Expressions containing the product of two vector-valued "
+                             "subexpressions cannot be used for assignment. Consider using "
+                             "interpolate instead.")
+        scaling = self._as_scalar(scalar)
+        return tuple((coeff, weight*scaling) for coeff, weight in vector)
+
+    def division(self, o, a, b):
+        # Division is only valid if b (the divisor) is a scalar
+        if self._is_scalar_equiv(b):
+            divisor = self._as_scalar(b)
+            return tuple((coeff, weight/divisor) for coeff, weight in a)
+        else:
+            raise ValueError("Expressions involving division by a vector-valued subexpression "
+                             "cannot be used for assignment. Consider using interpolate instead.")
+
+    def sum(self, o, a, b):
+        # Note: a and b are tuples of (coefficient, weight) so addition is concatenation
+        return a + b
+
+    def power(self, o, a, b):
+        # Only valid if a and b are scalars
+        return ((Constant(self._as_scalar(a) ** self._as_scalar(b)), 1),)
+
+    def abs(self, o, a):
+        # Only valid if a is a scalar
+        return ((Constant(abs(self._as_scalar(a))), 1),)
+
+    def _scalar(self, o):
+        return ((Constant(o), 1),)
+
+    int_value = _scalar
+    float_value = _scalar
+    zero = _scalar
+
+    def multi_index(self, o):
+        pass
+
+    def indexed(self, o, a, _):
+        return a
+
+    def component_tensor(self, o, a, _):
+        return a
+
+    def coefficient(self, o):
+        return ((o, 1),)
+
+    def expr(self, o, *operands):
+        raise NotImplementedError(f"Handler not defined for {type(o)}")
+
+    def _is_scalar_equiv(self, weighted_coefficients):
+        """Return ``True`` if the sequence of ``(coefficient, weight)`` can be compressed to
+        a single scalar value.
+
+        This is only true when all coefficients are :class:`firedrake.Constant` and have
+        shape ``(1,)``.
+        """
+        return all(isinstance(c, Constant) and c.dat.dim == (1,)
+                   for (c, _) in weighted_coefficients)
+
+    def _as_scalar(self, weighted_coefficients):
+        """Compress a sequence of ``(coefficient, weight)`` tuples to a single scalar value.
+
+        This is necessary because we do not know a priori whether a :class:`firedrake.Constant`
+        is going to be used as a scale factor (e.g. ``u.assign(Constant(2)*v)``), or as a
+        constant to be added (e.g. ``u.assign(2*v + Constant(3))``). Therefore we only
+        compress to a scalar when we know it is required (e.g. inside a product with a
+        :class:`firedrake.Function`).
+        """
+        return pytools.one(
+            functools.reduce(operator.add, (c.dat.data_ro*w for c, w in weighted_coefficients))
+        )
+
+
+# FIXME: This class is not annotated for the adjoint so taping will not work as expected.
```

Comment:
Fix this?

Reply:
I argue that this is a separate issue to resolve. Current adjoint approaches will still work fine. I've just added an extra entry point for `assign` that lacks the functionality.